src/material_pool.hpp

```diff
@@ -37,130 +52,83 @@ namespace wr
-				 bool double_sided = false);
+				 float alpha_constant = false,
+				 float double_sided = false);
 
 		Material(const Material& rhs) = default;
 
 		~Material();
 
-		TextureHandle GetAlbedo() { return m_albedo; }
-		void SetAlbedo(TextureHandle albedo);
-
-		void UseAlbedoTexture(bool use_albedo);
-
-		TextureHandle GetNormal() { return m_normal; }
-		void SetNormal(TextureHandle normal);
+		TextureHandle GetTexture(MaterialTextureType type);
+		void SetTexture(MaterialTextureType type, TextureHandle handle);
+		void ClearTexture(MaterialTextureType type);
+		bool HasTexture(MaterialTextureType type);
 
-		void UseNormalTexture(bool use_normal);
+		template<uint16_t s>
+		void GetConstant(MaterialConstantType type, float(&val)[s]);
+		
+		template<uint16_t s>
+		void SetConstant(MaterialConstantType type, float(&val)[s]);
 
-		TextureHandle GetRoughness() { return m_rougness; }
-		void SetRoughness(TextureHandle roughness);
+		void SetConstant(MaterialConstantType type, float val);
+		void GetConstant(MaterialConstantType type, float &val);
 
-		void UseRoughnessTexture(bool use_roughness);
-
-		TextureHandle GetMetallic() { return m_metallic; }
-		void SetMetallic(TextureHandle metallic);
-
-		void UseMetallicTexture(bool use_metallic);
+		TexturePool* const GetTexturePool() { return m_texture_pool; }
 
-		TextureHandle GetAmbientOcclusion() { return m_ao; }
-		void SetAmbientOcclusion(TextureHandle ao);
+		ConstantBufferHandle* GetConstantBufferHandle() { return m_constant_buffer_handle; };
+		void SetConstantBufferHandle(ConstantBufferHandle* handle) { m_constant_buffer_handle = handle; };
 
-		void UseAOTexture(bool use_ao);
+		void UpdateConstantBuffer();
 
-		DirectX::XMFLOAT3 GetConstantAlbedo()
+		union TextureFlags
 		{
-			return
-			{
-				DirectX::XMVectorGetX(m_material_data.m_color),
-				DirectX::XMVectorGetY(m_material_data.m_color),
-				DirectX::XMVectorGetZ(m_material_data.m_color)
+			struct {
+				uint32_t m_has_albedo_texture : 1;
+				uint32_t m_has_normal_texture : 1;
+				uint32_t m_has_roughness_texture : 1;
+				uint32_t m_has_metallic_texture : 1;
+				uint32_t m_has_ao_texture : 1;
 			};
-		};
-
-		void SetConstantAlbedo(DirectX::XMFLOAT3 color);
-
-		bool UsesConstantAlbedo() { return m_material_data.m_material_flags.m_use_albedo_constant; };
-		void SetUseConstantAlbedo(bool use_constant_albedo) { m_material_data.m_material_flags.m_use_albedo_constant = use_constant_albedo; };
 
-		float GetConstantAlpha() { return DirectX::XMVectorGetW(m_material_data.m_color); };
-		void SetConstantAlpha(float alpha);
-
-		DirectX::XMFLOAT3 GetConstantMetallic()
-		{
-			return
-			{
-				DirectX::XMVectorGetX(m_material_data.m_metallic_roughness),
-				DirectX::XMVectorGetY(m_material_data.m_metallic_roughness),
-				DirectX::XMVectorGetZ(m_material_data.m_metallic_roughness)
-			};
+			uint32_t m_value;
 		};
 
-		void SetConstantMetallic(DirectX::XMFLOAT3 metallic);
+		union MaterialData
```

Comment:
Does this mean we can't have an albedo texture and constants for the rest?

Reply:
Nope, it should still work exactly the same, just refactored and 'setConstant' and 'setTexture' are now used. You can just have an albedo texture and constants for everything else, I've tested it with pure constants and a mix and it worked fine.